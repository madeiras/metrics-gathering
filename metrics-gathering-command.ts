import { exec } from 'child_process';
import { promisify } from 'util';
import { appendFile, copyFile, readFile, stat, unlink, writeFile } from 'fs';

/**
 * Promisified commands
 */

export const appendFileCommand = promisify(appendFile);
export const copyFileCommand = promisify(copyFile);
export const executeCommand = promisify(exec);
export const readFileCommand = promisify(readFile);
export const removeFileCommand = promisify(unlink);
export const statCommand = promisify(stat);
export const writeFileCommand = promisify(writeFile);

/**
 * Get directory log
 */

export const getDirectoryLog = async ({
  limit,
  repositoryPath,
  skip,
  targetDirectory,
}: {
  limit: number;
  repositoryPath: string;
  skip: number;
  targetDirectory: string;
}) => {
  return executeCommand(
    `git --git-dir=${repositoryPath}/.git log --max-count=${limit} --pretty=format:"contributor:%an" --skip=${skip} --dirstat=files -- ${targetDirectory}`
  );
};

/**
 * Is file
 */

export const isFile = async ({ path }: { path: string }) => {
  try {
    await statCommand(path);
  } catch (error) {
    return false;
  }

  return true;
};

/**
 * Run
 */

export const run = async ({ repositoryPath, targetDirectory = 'packages' }: { repositoryPath: string; targetDirectory?: string }) => {
  if (typeof repositoryPath !== 'string') {
    throw new Error(`Unexpected input, expected "string", got "${typeof repositoryPath}"`);
  }

  try {
    await executeCommand(`git --git-dir=${repositoryPath}/.git rev-parse`);
  } catch (error) {
    throw new Error(`Failed to load repository at "${repositoryPath}" as a valid git repository`);
  }

  const contributorMap = {};
  const limit = 5000;
  let currentContributor;
  let directoryLogOutput = '';
  let skip = 0;

  console.log(`Starting evaluation of repository at: "${repositoryPath}"...`);

  do {
    const directoryLog = await getDirectoryLog({ limit, repositoryPath, skip, targetDirectory });

    if (directoryLog.stderr !== '') {
      throw new Error(`Unexpected error when retrieving dirstat data from repository (limit: "${limit}", skip: "${skip}"): "${directoryLog.stderr}"`);
    }

    directoryLogOutput = directoryLog.stdout;
    const logLines = directoryLogOutput.split('\n');

    // Parse the log lines
    for (const line of logLines) {
      if (line.trim() === '') {
        continue;
      }

      const [, contributor] = line.split('contributor:');

      if (contributor) {
        currentContributor = contributor;

        if (!contributorMap[contributor]) {
          contributorMap[contributor] = new Set();
        }

        continue;
      }

      // If this one is already guaranteed then skip
      if (contributorMap[currentContributor].size > 1) {
        continue;
      }

      const [, directory] = line.split(targetDirectory);
      const projectSubFolder = directory.split('/')[1];

      contributorMap[currentContributor].add(projectSubFolder);
    }

    skip += limit;
  } while (directoryLogOutput !== '');

  const result = Object.keys(contributorMap).filter((contributor) => contributorMap[contributor].size > 1).length;

  console.log(`Finished evaluation of repository at: "${repositoryPath}"`);
  console.log(`Result: "${result}"`);
  console.log('Writing result on README.md...');

  await writeResult({ repositoryPath, result });

  console.log('Finished writing result on README.md');
};

/**
 * Write result
 */

export const writeResult = async ({ repositoryPath, result }) => {
  const fileExists = await isFile({ path: './README.md' });

  if (!fileExists) {
    await copyFileCommand('./README.md.dist', './README.md');
    await appendFileCommand('./README.md', `\n\n## Latest result evaluating repository at "${repositoryPath}": ${result}\n`);

    return;
  }

  console.log('README.md already exists, checking file...');

  const fileData = await readFileCommand('./README.md', 'utf-8');
  const lines = fileData.split('\n');

  for (let index = 0; index < lines.length; index++) {
    if (lines[index].startsWith('## Latest result evaluating repository at "')) {
      console.log('Found result from last run on README.md, replacing...');

      lines[index] = `## Latest result evaluating repository at "${repositoryPath}": ${result}`;

      await writeFileCommand('./README.md', lines.join('\n'));

      return;
    }
  }

  console.log('Could not find result from last run on README.md, appending...');

  await appendFileCommand('./README.md', `\n\n## Latest result evaluating repository at "${repositoryPath}": ${result}\n`);
};
