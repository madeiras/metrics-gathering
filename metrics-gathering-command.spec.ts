import * as MethicsGatheringCommand from './metrics-gathering-command';

/**
 * Test `metrics-gathering`.
 */

describe('metrics-gathering', () => {
  afterAll(async () => {
    await MethicsGatheringCommand.copyFileCommand('./README.md.dist', './README.md');
  });

  beforeEach(async () => {
    const fileExists = await MethicsGatheringCommand.isFile({ path: './README.md' });

    if (fileExists) {
      await MethicsGatheringCommand.removeFileCommand('./README.md');
    }
  });

  describe('run()', () => {
    it('should throw an error if `repositoryPath` is not a string', async () => {
      try {
        await MethicsGatheringCommand.run({ repositoryPath: null as unknown as string });

        fail();
      } catch (error) {
        expect(error.message).toBe('Unexpected input, expected "string", got "object"');
      }
    });

    it('should throw an error if `repositoryPath` is not a valid git repo', async () => {
      try {
        await MethicsGatheringCommand.run({ repositoryPath: 'null' });

        fail();
      } catch (error) {
        expect(error.message).toBe('Failed to load repository at "null" as a valid git repository');
      }
    });

    it("should write the number of multiple contributors to a new README file if it doesn't exist", async () => {
      jest.spyOn(MethicsGatheringCommand, 'getDirectoryLog').mockReturnValueOnce(
        Promise.resolve({
          stdout:
            'contributor:Aliyah\n' +
            ' 100.0% packages/folderC/\n' +
            '\n' +
            'contributor:Bob\n' +
            '  50.0% packages/folderB/\n' +
            '  50.0% packages/folderC/\n' +
            '\n' +
            'contributor:Isaac Mann\n' +
            '  50.0% packages/folderA/\n' +
            '  50.0% packages/folderB/\n',
          stderr: '',
        })
      );
      jest.spyOn(MethicsGatheringCommand, 'getDirectoryLog').mockReturnValueOnce(Promise.resolve({ stdout: '', stderr: '' }));

      await MethicsGatheringCommand.run({ repositoryPath: '../count-contributors-sample' });

      const outputFile = await MethicsGatheringCommand.readFileCommand('./README.md', 'utf-8');

      expect(MethicsGatheringCommand.getDirectoryLog).toHaveBeenCalledTimes(2);
      expect(outputFile).toMatchSnapshot();
    });

    it('should write the number of multiple contributors to an existing README file with no result section by appending it', async () => {
      await MethicsGatheringCommand.appendFileCommand('./README.md', 'foobar');

      const existingFile = await MethicsGatheringCommand.readFileCommand('./README.md', 'utf-8');

      expect(existingFile).toMatchSnapshot();

      jest.spyOn(MethicsGatheringCommand, 'getDirectoryLog').mockReturnValueOnce(
        Promise.resolve({
          stdout:
            'contributor:Aliyah\n' +
            ' 100.0% packages/folderC/\n' +
            '\n' +
            'contributor:Bob\n' +
            '  50.0% packages/folderB/\n' +
            '  50.0% packages/folderC/\n' +
            '\n' +
            'contributor:Isaac Mann\n' +
            '  50.0% packages/folderA/\n' +
            '  50.0% packages/folderB/\n',
          stderr: '',
        })
      );
      jest.spyOn(MethicsGatheringCommand, 'getDirectoryLog').mockReturnValueOnce(Promise.resolve({ stdout: '', stderr: '' }));

      await MethicsGatheringCommand.run({ repositoryPath: '../count-contributors-sample' });

      const outputFile = await MethicsGatheringCommand.readFileCommand('./README.md', 'utf-8');

      expect(MethicsGatheringCommand.getDirectoryLog).toHaveBeenCalledTimes(2);
      expect(outputFile).toMatchSnapshot();
    });

    it('should write the number of multiple contributors to an existing README file with a result section by updating it', async () => {
      await MethicsGatheringCommand.appendFileCommand('./README.md', '\n\n## Latest result evaluating repository at "foobar": foobiz\n');

      const existingFile = await MethicsGatheringCommand.readFileCommand('./README.md', 'utf-8');

      expect(existingFile).toMatchSnapshot();

      jest.spyOn(MethicsGatheringCommand, 'getDirectoryLog').mockReturnValueOnce(
        Promise.resolve({
          stdout:
            'contributor:Aliyah\n' +
            ' 100.0% packages/folderC/\n' +
            '\n' +
            'contributor:Bob\n' +
            '  50.0% packages/folderB/\n' +
            '  50.0% packages/folderC/\n' +
            '\n' +
            'contributor:Isaac Mann\n' +
            '  50.0% packages/folderA/\n' +
            '  50.0% packages/folderB/\n',
          stderr: '',
        })
      );
      jest.spyOn(MethicsGatheringCommand, 'getDirectoryLog').mockReturnValueOnce(Promise.resolve({ stdout: '', stderr: '' }));

      await MethicsGatheringCommand.run({ repositoryPath: '../count-contributors-sample' });

      const outputFile = await MethicsGatheringCommand.readFileCommand('./README.md', 'utf-8');

      expect(MethicsGatheringCommand.getDirectoryLog).toHaveBeenCalledTimes(2);
      expect(outputFile).toMatchSnapshot();
    });
  });
});
