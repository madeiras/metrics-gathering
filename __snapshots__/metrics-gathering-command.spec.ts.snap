// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`metrics-gathering run() should write the number of multiple contributors to a new README file if it doesn't exist 1`] = `
"# metrics-gathering

A command to calculate the number of multiple contributors from a local repository that contains projects inside a \`packages\` folder, ideal for monorepos with this structure. 

## Installation

Install dependencies via \`yarn\`:

\`\`\`sh
❯ yarn
\`\`\`

or via \`npm\`:

\`\`\`sh
❯ npm install
\`\`\`

## Usage

The command can be run via \`index.ts\`:

\`\`\`sh
❯ ts-node ./index.ts --repositoryPath <absolute or relative path to a git repository>
\`\`\`

Additionally if you want to check a repository folder other than \`packages\`:

\`\`\`sh
❯ ts-node ./index.ts --repositoryPath <absolute or relative path to a git repository> --targetDirectory <packages is the default>
\`\`\`

__NOTE__: The main function is called \`run\` on \`metrics-gathering-command.ts\`.

## Tests

\`\`\`sh
❯ yarn test
\`\`\`

## License

MIT

## Latest result evaluating repository at "../count-contributors-sample": 2
"
`;

exports[`metrics-gathering run() should write the number of multiple contributors to an existing README file with a result section by updating it 1`] = `
"

## Latest result evaluating repository at "foobar": foobiz
"
`;

exports[`metrics-gathering run() should write the number of multiple contributors to an existing README file with a result section by updating it 2`] = `
"

## Latest result evaluating repository at "../count-contributors-sample": 2
"
`;

exports[`metrics-gathering run() should write the number of multiple contributors to an existing README file with no result section by appending it 1`] = `"foobar"`;

exports[`metrics-gathering run() should write the number of multiple contributors to an existing README file with no result section by appending it 2`] = `
"foobar

## Latest result evaluating repository at "../count-contributors-sample": 2
"
`;
